File: dp_labs_and_solutions/original/dp_labs_and_solutions/dp_java_labs_and_solutions/bridge/solution/RightAngledQuadrilateral.java
// RightAngledQuadrilateral.java

public class RightAngledQuadrilateral
{
	public RightAngledQuadrilateral(double aHeight, double aWidth)
	{
		myBody = makeBody(aHeight, aWidth);
	}

	public RightAngledQuadrilateral(double aSide)
	{
		myBody = makeBody(aSide, aSide);
		// or:
		myBody = new SquareBody(aSide);
	}

	public void setHeight(double aHeight)
	{
		myBody = makeBody(aHeight, getWidth());
	}

	public void setWidth(double aWidth)
	{
		myBody = makeBody(getHeight(), aWidth);
	}

	public void setSide(double aSide)
	{
		myBody = makeBody(aSide, aSide);
		// or:
		myBody = new SquareBody(aSide);
	}

	public double getHeight()
	{
		return myBody.getHeight();
	}

	public double getWidth()
	{
		return myBody.getWidth();
	}

	public boolean isSquare()
	{
		return myBody.isSquare();
	}

	private static RightAngledQuadrilateralBody makeBody(double aHeight, double aWidth)
	{
		if (aHeight == aWidth)
		{
			return new SquareBody(aHeight);
		}
		else
		{
			return new RectangleBody(aHeight, aWidth);
		}
	}

	private RightAngledQuadrilateralBody myBody;
}

File: dp_labs_and_solutions/dp_labs_and_solutions/dp_cpp_labs_and_solutions/bridge/lab/right_angled_quadrilateral.cpp
// right_angled_quadrilateral.cpp
#include <iostream>
#include "right_angled_quadrilateral.h"
#include "rectangle_body.h"
#include "square_body.h"

right_angled_quadrilateral::right_angled_quadrilateral(double a_height, double a_width)
{
    my_body = make_body(a_height, a_width);
}

right_angled_quadrilateral::right_angled_quadrilateral(double a_side)
{
    my_body = new square_body(a_side);
}


void right_angled_quadrilateral::set_height(double a_height)
{
    my_body = make_body(a_height, my_body->get_width());
}

void right_angled_quadrilateral::set_width(double a_width)
{
    my_body = make_body(my_body->get_height(), a_width);
}

void right_angled_quadrilateral::set_side(double a_side)
{
    my_body = make_body(a_side, a_side);
}

double right_angled_quadrilateral::get_height()
{
    return my_body->get_height();
}

double right_angled_quadrilateral::get_width()
{
    return my_body->get_width();
}

bool right_angled_quadrilateral::is_square()
{
    return my_body->is_square();
}

right_angled_quadrilateral_body* right_angled_quadrilateral::make_body(double a_height, double a_width)
{
    if (a_height <= 0 || a_width <= 0)
    {
        std::cerr << "Invalid parameters!" << std::endl;
        return nullptr;
    }
    if (a_height == a_width)
    {
        return new square_body(a_height);
    }
    else
    {
        return new rectangle_body(a_height, a_width);
    }

}
